File: scraper/scrapers/farmatodo.py
"""
Scraper para Farmatodo (farmatodo.com.co).
Estrategia: llama directamente al API de Algolia que usa el sitio internamente.
No requiere Playwright.
"""

import re
import httpx
from .base import BaseScraper, ScrapedProduct

ALGOLIA_URL = "https://api-search.farmatodo.com/1/indexes/products/query"
ALGOLIA_HEADERS = {
    "x-algolia-api-key": "eb9544fe7bfe7ec4c1aa5e5bf7740feb",
    "x-algolia-application-id": "VCOJEYD2PO",
    "content-type": "application/json",
}
HITS_PER_PAGE = 60


def _extract_active_ingredient(large_desc: str, description: str, sup_desc: str) -> str:
    if large_desc:
        m = re.search(r"Principio\s+Activo:\s*(.+?)(?:\s*\d|\s*$)", large_desc, re.IGNORECASE)
        if m:
            return m.group(1).strip()
    if sup_desc:
        parts = sup_desc.split()
        if parts:
            return parts[0].strip()
    # Fallback: primera palabra del nombre
    return description.split()[0] if description else ""


def _extract_concentration(description: str) -> str:
    m = re.search(r"(\d+(?:[.,]\d+)?\s*(?:mg|g|ml|mcg|ui|%|ug))", description, re.IGNORECASE)
    return m.group(1).strip() if m else ""


def _extract_presentation(description: str) -> str:
    forms = ["tabletas", "capsulas", "capsula", "tableta", "jarabe", "suspension",
             "solucion", "gotas", "ampolla", "supositorio", "crema", "gel",
             "pomada", "spray", "parche", "polvo", "inyectable"]
    desc_lower = description.lower()
    for form in forms:
        if form in desc_lower:
            return form.capitalize()
    return ""


def _classify(is_generic: bool, name: str) -> str:
    if is_generic:
        return "generic"
    name_lower = name.lower()
    generic_hints = ["genfar", " mk", "laproff", "procaps", "chalver",
                     "bussie", "coaspharma", "tecnoquimicas", "laboquimia",
                     "afidro", "audifarma"]
    for hint in generic_hints:
        if hint in name_lower:
            return "generic"
    return "brand"


def _parse_availability(without_stock: bool, stock: int) -> str:
    if without_stock:
        return "unavailable"
    if stock < 5:
        return "limited"
    return "available"


def _parse_discount(offer_text: str) -> int | None:
    if not offer_text:
        return None
    m = re.search(r"(\d+)", offer_text)
    return int(m.group(1)) if m else None


class FarmatodoScraper(BaseScraper):
    PHARMACY_ID = "farmatodo"

    async def search(self, query: str) -> list[ScrapedProduct]:
        results: list[ScrapedProduct] = []

        try:
            resp = httpx.post(
                ALGOLIA_URL,
                headers=ALGOLIA_HEADERS,
                json={
                    "query": query,
                    "hitsPerPage": HITS_PER_PAGE,
                    "filters": "categorie:'Salud y medicamentos' OR subCategory:'Drogueria'",
                },
                timeout=15,
            )
            data = resp.json()
        except Exception as e:
            print(f"[farmatodo] Error en API: {e}")
            return []

        hits = data.get("hits", [])
        for hit in hits:
            try:
                product = _map_hit(hit)
                if product:
                    results.append(product)
            except Exception:
                continue

        print(f"[farmatodo] '{query}' -> {len(results)} productos (API Algolia)")
        return results


def _map_hit(hit: dict) -> ScrapedProduct | None:
    name = hit.get("description") or hit.get("mediaDescription") or ""
    name = name.strip()
    if not name:
        return None

    full_price = hit.get("fullPrice") or 0
    offer_price = hit.get("offerPrice")

    # Precio final: offerPrice si existe y es valido, sino fullPrice
    if offer_price and offer_price > 0:
        price = int(offer_price)
    elif full_price and full_price > 0:
        price = int(full_price)
    else:
        return None

    # Precio mayor a 5M COP es un error de datos
    if price > 5_000_000:
        return None

    ref_price = int(full_price) if full_price and int(full_price) > price else None

    discount = _parse_discount(hit.get("offerText", ""))
    if ref_price and not discount:
        discount = round((1 - price / ref_price) * 100)

    large_desc = hit.get("large_description", "") or ""
    sup_desc = hit.get("sup_description", "") or ""
    ingredient = _extract_active_ingredient(large_desc, name, sup_desc)
    concentration = _extract_concentration(name)
    presentation = _extract_presentation(name)

    quantity_raw = hit.get("measurePum") or 1
    try:
        quantity = int(quantity_raw)
    except (ValueError, TypeError):
        quantity = 1

    price_per_unit = price // max(quantity, 1)

    without_stock = bool(hit.get("without_stock", True))
    stock = int(hit.get("stock") or 0)
    availability = _parse_availability(without_stock, stock)

    product_id = hit.get("id", "")
    url = f"https://www.farmatodo.com.co/product/{product_id}" if product_id else ""

    is_generic = bool(hit.get("isGeneric") or hit.get("generic"))

    return ScrapedProduct(
        pharmacy_id="farmatodo",
        product_name=name,
        type=_classify(is_generic, name),
        active_ingredient=ingredient,
        concentration=concentration,
        presentation=presentation,
        quantity=quantity,
        price=price,
        price_per_unit=price_per_unit,
        reference_price=ref_price,
        discount_pct=discount,
        availability=availability,
        url=url,
    )

File: app/lib/scrapers/farmatodo.ts
import type { ScrapedProduct } from './types'
import {
  extractConcentration,
  extractPresentation,
  extractPackQuantity,
  LIQUID_PRESENTATIONS,
  classify,
  matchesQuery,
} from './utils'
import { withCache } from './cache'

const ALGOLIA_URL = 'https://api-search.farmatodo.com/1/indexes/products/query'
const ALGOLIA_HEADERS = {
  'x-algolia-api-key':        'eb9544fe7bfe7ec4c1aa5e5bf7740feb',
  'x-algolia-application-id': 'VCOJEYD2PO',
  'content-type':             'application/json',
  'accept':                   'application/json',
}

// Fallback host if the custom proxy is down
const ALGOLIA_FALLBACK = 'https://VCOJEYD2PO-dsn.algolia.net/1/indexes/products/query'

function toSlug(str: string): string {
  return str
    .toLowerCase()
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '')
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function mapHit(hit: Record<string, any>): ScrapedProduct | null {
  const name = String(hit.description ?? hit.mediaDescription ?? '').trim()
  if (!name) return null

  const fullPrice  = Number(hit.fullPrice)  || 0
  // offerPrice in Algolia can be stale (expired promotions) — always use fullPrice
  // as the primary price to match what the user will actually pay on the site.
  const price = Math.round(fullPrice)
  if (price <= 0 || price > 5_000_000) return null

  let discount: number | undefined
  const offerText = String(hit.offerText ?? '').trim()
  const dm = offerText.match(/(\d+)/)
  if (dm) discount = parseInt(dm[1])

  const largDesc = String(hit.large_description ?? '')
  const supDesc  = String(hit.sup_description  ?? '')

  let ingredient = ''
  if (largDesc) {
    const m = largDesc.match(/Principio\s+Activo:\s*(.+?)(?:\s*\d|\s*$)/i)
    if (m) ingredient = m[1].trim()
  }
  if (!ingredient && supDesc) ingredient = supDesc.split(/\s/)[0]?.trim() ?? ''
  if (!ingredient) ingredient = name.split(/\s/)[0] ?? ''

  const presentation = extractPresentation(name)
  const pumQty = Number(hit.measurePum) || 0
  if (hit.measurePum == null && !LIQUID_PRESENTATIONS.has(presentation) && presentation !== '') {
    return null
  }
  const quantity = pumQty >= 2 ? pumQty : extractPackQuantity(name, presentation)

  const withoutStock = Boolean(hit.without_stock ?? false)
  const stock        = parseInt(String(hit.stock ?? 0)) || 0
  let availability: ScrapedProduct['availability'] = 'available'
  if (withoutStock) availability = 'unavailable'
  else if (stock > 0 && stock < 5) availability = 'limited'

  const productId = String(hit.id ?? '')
  const imageUrl  = String(hit.mediaImageUrl ?? hit.imageURL ?? '')

  // hit.url contains the pre-built slug: "{id}-{slug}" (e.g. "1004009-ibuprofeno-genfar-800mg-...").
  // Fall back to generating it from the product name when the field is absent.
  const urlSlug = hit.url
    ? String(hit.url)
    : productId
    ? `${productId}-${toSlug(name)}`
    : ''

  return {
    pharmacyId:       'farmatodo',
    productName:      name,
    type:             classify({ name, isGeneric: Boolean(hit.isGeneric || hit.generic) }),
    activeIngredient: ingredient,
    concentration:    extractConcentration(name, presentation),
    presentation,
    quantity:         Math.max(quantity, 1),
    price,
    pricePerUnit:     Math.round(price / Math.max(quantity, 1)),
    referencePrice:   undefined,
    discountPct:      discount,
    availability,
    url:      urlSlug ? `https://www.farmatodo.com.co/producto/${urlSlug}` : '',
    imageUrl: imageUrl || undefined,
  }
}

async function queryAlgolia(url: string, query: string): Promise<ScrapedProduct[] | null> {
  const res = await fetch(url, {
    method:  'POST',
    headers: ALGOLIA_HEADERS,
    body: JSON.stringify({
      query,
      hitsPerPage: 60,
      filters: "categorie:'Salud y medicamentos' OR subCategory:'Drogueria'",
    }),
    signal: AbortSignal.timeout(7_000),
  })
  if (!res.ok) return null
  const data = await res.json() as { hits?: Record<string, unknown>[] }
  if (!Array.isArray(data.hits)) return null
  return data.hits.flatMap(h => {
    const p = mapHit(h as Record<string, unknown>)
    return p ? [p] : []
  })
}

export async function searchFarmatodo(query: string): Promise<ScrapedProduct[]> {
  return withCache('farmatodo', query, async () => {
    try {
      // Try primary host first, fall back to standard Algolia DSN
      let hits = await queryAlgolia(ALGOLIA_URL, query)
      if (!hits) hits = await queryAlgolia(ALGOLIA_FALLBACK, query)
      if (!hits) {
        console.error('[farmatodo] Both Algolia hosts failed')
        return null // failure -> serve stale cache
      }

      return hits.filter(r => matchesQuery(query, r.productName, r.activeIngredient))
    } catch (e) {
      console.error('[farmatodo] Error:', e)
      return null
    }
  })
}
